### ISSUE TITLE
OptionSet for diff not called with :diffsplit and "set diff" during VimEnter

### ISSUE THREAD

**Issue by blueyed** [vim/vim#2544]
Given `t-optionset-diff.vim` calling `vim -Nu t-optionset-diff.vim` will echo 0 twice.
Calling `:call Test()` then will echo 0 and 1.

I.e. it does not get called with `:diffsplit` (known issue?! - see https://github.com/vim/vim/issues/1286), but also not before/within VimEnter.
The VimEnter issue might be by design, but then the help is misleading, since it looks like you could use it to create dirs automatically based on your settings:

> Usage example: Check for the existence of the directory in the 'backupdir'
> and 'undodir' options, create the directory if it doesn't exist yet.

(This seems to only work when setting it manually / after vimrc is loaded?!)

```vim
let g:diff_set = 0

function VimEnter()
  augroup test
    autocmd OptionSet diff let g:diff_set += 1
  augroup END

  call Test()
endfunction

function Test()
  diffsplit
  echom g:diff_set
  set diff
  echom g:diff_set
endfunction

augroup test
  autocmd VimEnter * call VimEnter()
augroup END
```

**Comment by chrisbra** [comment]
Yes, it is a known issue, that OptionSet is not called for some of the diff options, because internally only some state variables are set and not the `set_..._option..` functions are called. Second, I believe, that this is mentioned in the documentation, that OptionSet is not called during startup to not disturb loading too much. Perhaps we need to adjust the help file then.

**Comment by brammool** [comment]
Daniel Hahler wrote:

> Given `t-optionset-diff.vim` calling `vim -Nu t-optionset-diff.vim` will echo 0 twice.
> Calling `:call Test()` then will echo 0 and 1.
> 
> I.e. it does not get called with `:diffsplit` (known issue?! - see https://github.com/vim/vim/issues/1286), but also not before/within VimEnter.
> The VimEnter issue might be by design, but then the help is misleading, since it looks like you could use it to create dirs automatically based on your settings:
> 
> > 				Usage example: Check for the existence of the
> 				directory in the 'backupdir' and 'undodir'
> 				options, create the directory if it doesn't
> 				exist yet.
> 
> (This seems to only work when setting it manually / after vimrc is loaded?!)
> 
> ```vim
> let g:diff_set = 0
> 
> function VimEnter()
>   augroup test
>     autocmd OptionSet diff let g:diff_set += 1
>   augroup END
> 
>   call Test()
> endfunction
> 
> function Test()
>   diffsplit

This requires an argument.  Diffing a file with itself is pointless.

>   echom g:diff_set
>   set diff
>   echom g:diff_set
> endfunction
> 
> augroup test
>   autocmd VimEnter * call VimEnter()
> augroup END
> ```

When I use:
	:let g:diff_set = 0
	:au OptionSet diff let g:diff_set += 1
	:diffsplit somefile

Then g:diff_set is 2 (once incremented for each file).  That looks
correct.

Not sure about what is supposed to happen with VimEnter, I don't see why
you would only install the autocommand then.



-- 
hundred-and-one symptoms of being an internet addict:
264. You turn to the teletext page "surfing report" and are surprised that it
     is about sizes of waves and a weather forecast for seaside resorts.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///
